mne_sandbox/preprocessing/_sns.py

```diff
@@ -0,0 +1,116 @@
+    logger.info('    Loading raw data')
+    if not isinstance(raw, _BaseRaw):
+        raise TypeError('raw must be an instance of Raw, got %s' % type(raw))
+    n_neighbors = int(n_neighbors)
+    if n_neighbors < 1:
+        raise ValueError('n_neighbors must be positive')
+    good_picks = _pick_data_channels(raw.info, exclude='bads')
+    if n_neighbors > len(good_picks) - 1:
+        raise ValueError('n_neighbors must be at most len(good_picks) - 1 (%s)'
+                         % (len(good_picks) - 1,))
+    logger.info('    Loading data')
+    raw = raw.copy()
+    raw.load_data()
+    picks = _pick_data_channels(raw.info, exclude=())
+    bad_picks = np.setdiff1d(picks, good_picks)
+    # The following few lines are equivalent to this, but require less mem use:
+    # data_cov = np.cov(orig_data)
+    # data_corrs = np.corrcoef(orig_data) ** 2)
+    logger.info('    Computing covariance for %s good channels'
+                % len(good_picks))
+    data_cov = np.eye(len(picks))
+    data_cov[good_picks[:, np.newaxis], good_picks] = compute_raw_covariance(
+        raw, picks=good_picks, reject=reject, flat=flat, verbose=False)['data']
+    del good_picks
+    data_norm = np.diag(data_cov)
+    data_corrs = data_cov * data_cov
+    data_corrs /= data_norm
+    data_corrs /= data_norm[:, np.newaxis]
+    del data_norm
+    data_cov *= len(raw.times)
+    operator = np.zeros((len(picks), len(picks)))
+    logger.info('    Assembling spatial operator')
+    for ii in range(len(picks)):
+        # For each channel, the set of other signals is orthogonalized by
+        # applying PCA to obtain an orthogonal basis of the subspace spanned
+        # by the other channels.
+        if picks[ii] in bad_picks:
+            operator[ii, ii] = 1.
+            continue
+        idx = np.argsort(data_corrs[ii])[::-1][:n_neighbors + 1].tolist()
+        idx.pop(idx.index(ii))  # should be in there
+        # XXX Eventually we might want to actually threshold here (with
+        # rank-deficient data it could matter)
+        s, v = _pca(data_cov[idx][:, idx], thresh=None)
+        v *= 1. / np.sqrt(s)
+        # augment with given channel
+        v = np.vstack(([[1] + [0] * n_neighbors],
+                       np.hstack((np.zeros((n_neighbors, 1)), v))))
+        idx = np.concatenate(([ii], idx))
+        corr = np.dot(np.dot(v.T, data_cov[idx][:, idx]), v)
+        # The channel is projected on this basis and replaced by its projection
+        operator[ii, idx[1:]] = np.dot(corr[0, 1:], v[1:, 1:].T)
+    offsets = np.concatenate([np.arange(0, len(raw.times), 10000),
+                              [len(raw.times)]])
+    logger.info('    Applying operator')
+    for start, stop in zip(offsets[:-1], offsets[1:]):
+        time_sl = slice(start, stop)
+        raw._data[picks, time_sl] = np.dot(operator, raw._data[picks, time_sl])
+    logger.info('Done')
+    return raw, operator
```

Comment:
how about you make a class that stores the operator matrix as attribute so you can call apply on any instance? so we converge with the same API as ICA and ideally DSS